_posts/pl/2023-10-13-jak-zmusic-mongodb-do-uzycia-indeksu-bez-zmiany-kodu.md

```diff
@@ -0,0 +1,1050 @@
+  },
+  "ok": 1,
+  "operationTime": Timestamp(1671465640, 1),
+  "$clusterTime": {
+    "clusterTime": Timestamp(1671465640, 1),
+    "signature": {
+      "hash": BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
+      "keyId": NumberLong(0)
+    }
+  }
+}
+```
+Wynik analizy jest dość obszerny jednak można go podsumować tak:
+- `winningPlan` czyli plan, który został wykonany posłużył się standardowym indeksem kolekcji:
+```javascript
+  {
+    "inputStage": {
+      "stage": "IXSCAN",
+      "keyPattern": {
+        "_id": 1
+      },
+      "indexName": "_id_",
+      "isMultiKey": false,
+      "multiKeyPaths": {
+        "_id": [ ]
+      },
+      "isUnique": true,
+      "isSparse": false,
+      "isPartial": false,
+      "indexVersion": 2,
+      "direction": "backward",
+      "indexBounds": {
+        "_id": [
+          "[MaxKey, MinKey]
+        ]
+      }
+    }
+  }
+```        
+- `rejectedPlans` czyli plany odrzucone zawierają plan wykorzystujący indeks, który naszym zdaniem powinien zostać użyty:
+```javascript
+  {
+    "inputStage": {
+      "stage": "IXSCAN",
+      "keyPattern": {
+        "formFields.formInstanceNumber.value": 1
+      },
+      "indexName": "formFields.formInstanceNumber.value_1",
+      "isMultiKey": false,
+      "multiKeyPaths": {
+        "formFields.formInstanceNumber.value": [ ]
+      },
+      "isUnique": false,
+      "isSparse": false,
+      "isPartial": false,
+      "indexVersion": 2,
+      "direction": "forward"
+    }
+  }
+```
+
+Co więcej, plan który został wybrany odwiedził 254851 dokumentów i kluczy, a zapytanie trwało 11651 ms:
+```javascript
+  {
+    "executionStats": {
+      "executionSuccess": true,
+      "nReturned": 75,
+      "executionTimeMillis": 11651,
+      "totalKeysExamined": 254851,
+      "totalDocsExamined": 254851
+    }
+  }
+```
+Przy czym jeden z planów odrzuconych mógł zwrócić wynik po odwiedzeniu 75 dokumentów i 135 kluczy, a jego wykonanie szacowane było na 30 ms:
+```javascript
+{
+  "nReturned": 0,
+  "executionTimeMillisEstimate": 30,
+  "totalKeysExamined": 135,
+  "totalDocsExamined": 75
+}
+```
+## Dlaczego MongoDB nie używa indeksu
+
+W tym miejscu należy się zastanowić w jaki sposób MongoDB wybiera najlepszy plan zapytania. Z pierwszego punktu tego wpisu wiemy, że optymalizator bazuje na statystykach zbieranych podczas działania bazy. Potrzebujemy jeszcze wiedzieć w jaki sposób budowane są te statystyki. W opisywanym przypadku to właśnie tutaj kryje się rozwiązanie naszego problemu.
+
+Optymalizator MongoDB cache'uje plany zapytań. Plan zapytania, który wygrał (`winningPlan`) trafia do cache'a i po kolejnym zapytaniu, w którym okazał się planem wygrywającym staje się aktywny. Następne zapytanie o takim samym **kształcie** zostaje wykonane w oparciu o aktywny plan z cache'a. Algorytm wygląda tak:
+
+![Algorytm optymalizatora](/assets/img/posts/2023-10-13-jak-zmusic-mongodb-do-uzycia-indeksu-bez-zmiany-kodu/query-planner-logic.bakedsvg.svg)
```

Comment:
Zakładam, że nie jesteś autorem tej grafiki. Dorzućmy proszę jej źródło pod spodem. Żeby wszystko się ładnie wyświetliło trzeba to zrobić w HTMLu:
```
<div class="img-with-legend">
<img alt="Algorytm optymalizatora" src="/assets/img/posts/2023-10-13-jak-zmusic-mongodb-do-uzycia-indeksu-bez-zmiany-kodu/query-planner-logic.bakedsvg.svg" />
<span class="img-legend">Algorytm optymalizatora<br />źródło: <a href="https://www.jakas.strona.com">www.jakas.strona.com</a> - dostęp: 2023-10-17</span>
</div>
```